lua/telescope/_extensions/marks.lua

```diff
@@ -1,135 +1,284 @@
+    local sorter = conf.generic_sorter(opts)
+    local generic_scoring = sorter.scoring_function
+    sorter.scoring_function = function(self, prompt, line, entry)
+        local score = generic_scoring(self, prompt, line, entry)
+        local multiplier = 1
 
-    local selection = action_state.get_selected_entry()
-    harpoon:list():remove(selection.value)
+        -- set the multiplier, when matching an index
+        local index = entry.index
+        local index_str = tostring(index)
+        for value in string.gmatch(prompt, "%S+") do
+            local num = tonumber(value)
+            if num ~= nil then
+                if num == index then
+                    multiplier = 0.25
+                    break -- found an exact match
+                elseif index_str:match(value) then
+                    multiplier = 0.5
+                    -- continue looking for a better match
+                end
+            end
+        end
 
-    local function get_selections()
-        local results = {}
-        action_utils.map_selections(prompt_bufnr, function(entry)
-            table.insert(results, entry)
-        end)
-        return results
+        if score ~= -1 then -- generic_sorter found a match
+            score = score * multiplier -- make the score better
+        elseif multiplier ~= 1 then -- has matched an index
+            score = multiplier -- make the score slightly better
+        end
+
+        return score
     end
+    return sorter
+end
+
+local delete_mark_selections = function(prompt_bufnr)
+    local selections = {}
+    action_utils.map_selections(prompt_bufnr, function(entry)
+        table.insert(selections, entry)
+    end)
+    table.sort(selections, function(a, b)
+        return a.index < b.index
+    end)
 
-    local selections = get_selections()
-    for _, current_selection in ipairs(selections) do
-        harpoon:list():remove(current_selection.value)
+    local count = 0
+
+    if #selections > 0 then
+        -- delete marks from multi-selection
+        for i = #selections, 1, -1 do
+            local selection = selections[i]
+            harpoon:list():removeAt(selection.index)
+            count = count + 1
+        end
+    else
+        -- delete marks from single-selection
+        local selection = action_state.get_selected_entry()
+        if selection ~= nil then
+            harpoon:list():removeAt(selection.index)
+            count = count + 1
+        else
+            return 0
+        end
     end
 
+    -- delete picker-selections
     local current_picker = action_state.get_current_picker(prompt_bufnr)
-    current_picker:refresh(generate_new_finder(), { reset_prompt = true })
+    current_picker:delete_selection(function() end)
+
+    return count
 end
 
-local move_mark_up = function(prompt_bufnr)
-    local selection = action_state.get_selected_entry()
-    local length = harpoon:list():length()
+local delete_mark_selections_prompt = function(prompt_bufnr)
```

Comment:
nitpick: it would be great if this could be configurable (i.e. being able to disable this prompt and delete the mark straight away)

Reply:
I could also imagine to override the mappings!? Something like this:

```lua
require("telescope").extensions.harpoon.marks(themes.get_dropdown({
    previewer = false,
    layout_config = { width = 0.6 },
    path_display = { truncate = 10 },
    attach_mappings = function(_, map)
        map("i", "<c-d>", hm_actions.delete_mark_selections)
        map("n", "<c-d>", hm_actions.delete_mark_selections)
        return true
    end,
}))
```

The different actions (e.g. `delete_mark_selections`, `delete_mark_selections_prompt`) would've to be exported first!?

For now, you could just override the mappings in the code and use `delete_mark_selections` instead of `delete_mark_selections_prompt`!

Reply:
That would be even better! :)

Indeed, the actions need to be exported first. I'm using your fork now, thanks for doing this again 🙏🏽

Reply:
You can now do:

```lua
local themes = require("telescope.themes")
local hm_actions = require("telescope._extensions.harpoon_marks.actions")
require("telescope").extensions.harpoon.marks(themes.get_dropdown({
    previewer = false,
    layout_config = { width = 0.6 },
    path_display = { truncate = 10 },
    attach_mappings = function(_, map)
        map("i", "<c-d>", hm_actions.delete_mark_selections)
        map("n", "<c-d>", hm_actions.delete_mark_selections)
        return true
    end,
}))
```